### PR TITLE
Feature/event bubbling

### LINKED ISSUE #164: UI Infrastructure: Event Bubbling
At the current time, we have a very simple event model for handling mouse events. 

That logic is in https://github.com/revery-ui/revery/blob/master/src/UI/Mouse.re , specifically here:
```
    let isNodeImpacted = n => n#hitTest(pos);
    let nodes: ref(list(Node.node('a))) = ref([]);
    let collect = n =>
      if (isNodeImpacted(n)) {
        nodes := List.append(nodes^, [n]);
      };
    Node.iter(collect, node);
    List.iter(n => n#handleEvent(eventToSend), nodes^);
```

This is _very simple_ - we check for all the nodes that pass the 'hit-test', and dispatch the event to all of them.

However, this is unexpected behavior and not intuitive if you're coming from web programming - as webdevs, we'd expect the event to dispatch to the _top-most_ element, and bubble up from there!

We need to implement this event-bubbling behavior in revery. This will be useful not just for the initial mouse events, but for all sorts of other events - like keyboard input, etc.

## Proposal

We add a `UiEvents` module that has a method `bubble = (node, event)`. The `bubble` would do a few things:
- It would wrap the event with some extra methods, like `stopPropagation` or `preventDefault` - like we'd expect in Web events.
- It would keep some internal state of whether propagation was stopped or default was prevented (ie, via `ref`s).
- It would call `handleEvent` for each node in the hierarchy. If `stopPropagation` is called, it should discontinue the traversal up the hierarchy.

## Testability

We should be able to craft unit tests that exercise this with some simple `Node` objects that have handlers that call `stopPropagation`, along with counters that validate whether the events were hit.

## Application

We can hook this up to the mouse event bubbling behavior today, by picking out the _top-most node_ that passes the hit-test. We need to make sure our z-index tracking is working correctly for this.

### PR DESCRIPTION
**Pull request by akinsho**
This PR will eventually add event bubbling to revery and will fix #164.

### CODE REVIEW

**Comment by OhadRau** [comment]
Hmm, correct me if I'm wrong but I think `Mouse.dispatch` is always called with the root node, so I think this would fail to bubble since the root has no parent. You can see the way I'm doing the pseudo-bubbling in the cursor-styles PR here: https://github.com/OhadRau/revery/blob/cursor-styles/src/UI/Mouse.re#L100-L117 (i.e. find the deepest node in the hierarchy and then get it's cursor style property which more or less bubbles up).

The `getDepth` method is implemented through a `_depth` reference that gets calculated when a node's parent changes (https://github.com/OhadRau/revery/blob/cursor-styles/src/UI/Node.re#L134-L137) and then I implemented `getCursorStyle` like this: https://github.com/OhadRau/revery/blob/cursor-styles/src/UI/Node.re#L71-L77. I think this sort of approach can be used for traversing the node hierarchy here too, but I might've missed something in your changes.

File: src/UI/UiEvents.re [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+      | Some(parent) => traverseHeirarchy(parent, bubbled)
+      | None => ()
+      };
+    }
+  );
+
+let bubble = (node: node('a), event: mouseEvent) => {
+  /* Wrap event with preventDefault and stopPropagation */
+  let evt = BubbledEvent.make(event);
+  traverseHeirarchy(node, evt);
```
Comment (bryphe) [inline src/UI/UiEvents.re]: NIT: Spelling - should be `traverseHierarchy`

File: src/UI/UiEvents.re [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+      };
+  };
+};
+
+let rec traverseHeirarchy = (node: node('a), bubbled) =>
+  BubbledEvent.(
+    /* track if default prevent or propagation stopped per module
+       stop traversing node hierarchy if stop propagation is called
+        */
+    if (bubbled.shouldPropagate) {
```
Comment (bryphe) [inline src/UI/UiEvents.re]: Logic looks great here - thanks for implementing the bubbling 👍

File: src/UI/UiEvents.re [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    let alreadyExists =
+      List.fold_left(
+        (found, elem) => found || elem.id == id,
+        false,
+        events^,
+      );
+    alreadyExists ?
+      wrappedEvent :
+      {
+        events := List.append([], [wrappedEvent]);
```
Comment (bryphe) [inline src/UI/UiEvents.re]: It seems like we're always adding to the `events` list - but I didn't see where events are getting removed? It wasn't quite clear to me why we need a list of the events.

If we only care about the 'active' event - would it make sense to use a ref like `let activeEvent = ref(event)`?

File: src/UI/UiEvents.re [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+      List.map(
+        evt => evt.id == id ? {...evt, defaultPrevented: true} : evt,
+        events^,
+      );
+    List.find(evt => evt.id == id, events^);
+  };
+
+  let make = event => {
+    let id = getId();
+    let wrappedEvent = {
```
Comment (bryphe) [inline src/UI/UiEvents.re]: If `shouldPropagate` and `defaultPrevented` were defined either as `ref(bool)`, we could keep the implementation of `stopPropagation` and `preventDefault` local to the event, like:
```
let shouldPropagate = ref(true);
let defaultPrevented = ref(false);

let stopPropagation = () => shouldPropagate := false;
let preventDefault = () => defaultedPrevented := true;

let wrappedEvent = {
  id,
  event,
  shouldPropagate,
  defaultPrevented, 
  stopPropagation,
  preventDefault,
}; 
```

But perhaps I'm missing an advantage of them being in a list?

File: src/UI/UiEvents.re (hunk 1 above)
Comment (Akin909) [reply src/UI/UiEvents.re]: :+1:

File: test/UI/MouseTest.re [hunk 5]
```diff
@@ -48,6 +49,26 @@ test("Mouse", () => {
...
     });
   });
+
+  test("bubbleEvent", () =>
+    test(
+      "test that state is updated per event when stop propagation is called",
+      () => {
+      let cursor = Mouse.Cursor.make();
+      let testEvent =
+        Mouse.internalToExternalEvent(
```
[the remark this answers is not in the document]
Comment (Akin909) [reply test/UI/MouseTest.re]: Thanks, I've tweaked this now to just use the `MouseMove`

File: src/UI/Mouse.re [hunk 6]
```diff
@@ -93,6 +94,7 @@ let dispatch =
   let pos = getPositionFromMouseEvent(cursor, evt);
 
   let eventToSend = internalToExternalEvent(cursor, evt);
+  bubble(node, eventToSend);
```
Comment (bryphe) [inline src/UI/Mouse.re]: I think this should actually be inside the `if(!handleCapture(..))` block.  The reason is - when a capture is active, it takes priority over other events. It gets 'first dibs' on handling the event, in other words. Helpful for a bunch of scenarios!

Could we replace that entire block inside with your new `bubble` function? The previous strategy collected all the nodes, and dispatched against the deepest node - but I think your `bubble` strategy supercedes now.

At the very least - I think we could replace:
```
    List.iter(n => n#handleEvent(eventToSend), nodes^);
```
with the new `bubble` function.

It seems like we still need some logic there to set the proper cursor style - but I think we could move `bubble` in to replace that `n#handleEvent` call.

File: src/UI/UiEvents.re (hunk 3 above)
Comment (bryphe) [reply src/UI/UiEvents.re]: Cool! I like the new approach better, thank you for making the change!

File: test/UI/MouseTest.re (hunk 5 above)
[the remark this answers is not in the document]
Comment (bryphe) [reply test/UI/MouseTest.re]: Excellent, thanks!

**Pull request merged**